Submit STAC using transactions API

ividito opened the pull request:
Proposed implementation for #288 

Original ingest API support is maintained through an Airflow variable. I learned while setting this up that this is [not a recommended approach](https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#dynamic-dags-with-environment-variables), but it is consistent with what we do elsewhere.

ividito committed:
feat: Use stac bulk transactions

ividito committed:
Adjust transactions variable

amarouane-ABDELHAK commented on dags/veda_data_pipeline/utils/submit_stac_transactions.py:
```diff
@@ -0,0 +1,121 @@
+class TransactionsApi:
+
+    @classmethod
+    def from_veda_auth_secret(cls, *, secret_id: str, base_url: str) -> "IngestionApi":
```
Shouldn't the return type hint be "TransactionsApi" instead of "IngestionApi"

amarouane-ABDELHAK commented on dags/veda_data_pipeline/utils/submit_stac_transactions.py:
```diff
@@ -0,0 +1,121 @@
+            response.raise_for_status()
+        except Exception as ex:
+            print(response.text)
+            raise f"Error, {ex}"
```
Raise should be derived from a baseException otherwise it won't raise a string 🤔
Probably we should use
```python
raise RuntimeError(f"Error, {ex}")

```

amarouane-ABDELHAK commented on dags/veda_data_pipeline/utils/submit_stac_transactions.py:
```diff
@@ -0,0 +1,121 @@
+        :param token: Optional Bearer token for authenticated STAC APIs.
+        """
+        self.stac_ingestor_api_url = stac_ingestor_api_url.rstrip('/')
+        self.cognito_app_secret = cognito_app_secret
```
Where are we using `self.cognito_app_secret`?

amarouane-ABDELHAK reviewed (commented):
Thank you for the PR. I just left some small comments

ividito committed:
PR review feedback
- corrected type hint
- corrected exception in auth method
- removed unused cognito var

smohiudd committed:
refactor submit stac transactions

smohiudd commented:
This branch has been deployed to SIT, but testing is currently blocked because the dev STAC API has switched over to using veda-keycloak and we still need to sort out how programatic auth will work.

smohiudd committed:
use keycloak for transactionss POST

smohiudd committed:
update stac transactions util

smohiudd committed:
fix tf transactions boolean

smohiudd commented:
This branch was tested by switching on and off the `TF_VAR_transactions_endpoint_enabled` feature flag. I ingested the `ida-ndwi-difference-test-transactions` collection into the [dev catalog](https://dev.openveda.cloud/stac/collections/ida-ndwi-difference-test-transactions).

smohiudd commented on sm2a/infrastructure/main.tf:
```diff
@@ -104,7 +104,9 @@ module "sma-base" {
     CLOUDFRONT_TO_INVALIDATE = var.cloudfront_to_invalidate
-    CLOUDFRONT_PATH_TO_INVALIDATE = var.cloudfront_path_to_invalidate
+    CLOUDFRONT_PATH_TO_INVALIDATE = var.cloudfront_path_to_invalidate,
+    TRANSACTIONS_ENDPOINT_ENABLED = var.transactions_endpoint_enabled==true ? "True" : null
```
I don't think tf bools are serialized in airflow in python. I used this hack but open if there's a better way.

smohiudd committed:
add back success_file exception

smohiudd committed:
remove success_file duplication

smohiudd committed:
fix main.tf

smohiudd committed:
add prefix to lambda trigger role

botanical commented on dags/veda_data_pipeline/groups/processing_tasks.py:
```diff
@@ -1,14 +1,17 @@
 
 group_kwgs = {"group_id": "Process", "tooltip": "Process"}
 
+airflow_vars = Variable.get("aws_dags_variables")
```
If we set deserialize_json=True, do you know if we still need to do:
```
var.transactions_endpoint_enabled==true ? "True" : null
```
in main.tf? 

```suggestion
airflow_vars = Variable.get("aws_dags_variables", deserialize_json=True)
```

And maybe we refactor this file a bit so we call Variable.get in tasks rather than in the top level, since the docs seem to imply this is best practice: https://airflow.apache.org/docs/apache-airflow/2.8.4/best-practices.html#airflow-variables

ividito replied:
I would love a refactor for this (we call the same set of `Variable.get("aws_dags_variables")` lines 16 times in our DAGs), but I think that's a different issue. `.airflowignore` keeps the DAG processor from parsing the `groups` and `utils` folders, so there's no performance loss by having these calls here.

smohiudd replied:
I tried using `airflow_vars_json = Variable.get("aws_dags_variables", deserialize_json=True)` and I'm getting a decode error: `requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`

ividito commented on sm2a/infrastructure/main.tf:
```diff
@@ -103,8 +103,10 @@ module "sma-base" {
-    CLOUDFRONT_PATH_TO_INVALIDATE = var.cloudfront_path_to_invalidate
+    CLOUDFRONT_TO_INVALIDATE = var.cloudfront_to_invalidate,
+    CLOUDFRONT_PATH_TO_INVALIDATE = var.cloudfront_path_to_invalidate,
+    TRANSACTIONS_ENDPOINT_ENABLED = var.transactions_endpoint_enabled==true ? "True" : null,
```
Related fix that removes the need for the ternary here: https://github.com/NASA-IMPACT/self-managed-apache-airflow/pull/18

cc/ @amarouane-ABDELHAK

smohiudd committed:
use transactions PUT instead of bulk items

ividito commented:
^^ rebased to pull in changes from dev

smohiudd commented:
Based on conversations with stac-fastapi maintainers, they recommend we don't use bulk ingestion. I tried instead to use the PUT {collection_id}/items/{item_id} endpoint for create and update operations but stac-fastapi only supports [updates](https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/extensions/stac_fastapi/extensions/core/transaction.py#L107). 

The submit stac util will need to be refactored to handle both creating new items via POST and updating items via PUT. 

@ividito I suggest we proceed with mwaa refactor PR since this transactions PR will need more time.
